relay-general/src/store/transactions/processor.rs

```diff
@@ -1,21 +1,89 @@
+use chrono::Utc;
+
 use relay_common::SpanStatus;
 
 use crate::processor::{ProcessValue, ProcessingState, Processor};
 use crate::protocol::{
-    Context, ContextInner, Event, EventType, Span, Timestamp, TransactionSource,
+    Context, ContextInner, Event, EventType, Span, Timestamp, TransactionInfo, TransactionSource,
 };
 use crate::store::regexes::TRANSACTION_NAME_NORMALIZER_REGEX;
 use crate::types::{Annotated, Meta, ProcessingAction, ProcessingResult, Remark, RemarkType};
 
+use super::TransactionNameRule;
+
 /// Rejects transactions based on required fields.
 #[derive(Default)]
-pub struct TransactionsProcessor {
+pub struct TransactionsProcessor<'r> {
     normalize_names: bool,
+    tx_name_rules: &'r [TransactionNameRule],
 }
 
-impl TransactionsProcessor {
-    pub fn new(normalize_names: bool) -> Self {
-        Self { normalize_names }
+impl<'r> TransactionsProcessor<'r> {
+    pub fn new(normalize_names: bool, tx_name_rules: &'r [TransactionNameRule]) -> Self {
+        Self {
+            normalize_names,
+            tx_name_rules,
+        }
+    }
+
+    /// Applies the rule if any found to the transaction name.
+    ///
+    /// It find the first rule matching the criteria:
+    /// - source matchining the one provided in the rule sorce, default `url`
+    /// - rule hasn't epired yet
+    /// - glob pattern matches the transaction name
+    ///
+    /// Note: we add `/` at the end of the transaction name if there isn't one, to make sure that
+    /// patterns like `/<something>/*/**` where we have `**` at the end are a match.
+    pub fn apply_transaction_rename_rule(
+        &self,
+        transaction: &mut Annotated<String>,
+        info: &mut std::option::Option<TransactionInfo>,
+        source: &TransactionSource,
+    ) -> ProcessingResult {
+        let now = Utc::now();
+        transaction.apply(|transaction, meta| {
+            let slash_is_present = transaction
+                .chars()
+                .last()
+                .map(|c| c == '/')
+                .unwrap_or_default();
+
+            // Add new `/` at the end of the transaction if there isn't one.
+            if !slash_is_present {
+                transaction.push('/');
+            }
```

Comment:
It might make the code a little bit prettier in this one case but one will have to jump to different function to check what's going on. We could maybe come back to this in the followup PRs.

This also can be complicated to do, since we have few cases, if the rules applied we should remove the added `/` and save the results in to proper places, or another case when we clean up the added slash afterwards. 

If you could sketch out  how you envision the code of the `with_slashed_transaction` would look like? 

Should this function be added to `Event` protocol or keep it here and get event as the input data?

Reply:
My thought was to leave it here, but adding it to the `Event` might not be a bad idea either. My idea is the following, where `closure` is a function with the logic below:

```
def with_slashed_transaction(tx, closure):
  if slash_is_present(tx):
      closure()
  else:
      add_slash(tx)
      closure()
      remove_slash(tx)
```